pkg/pss/pss_test.go

```diff
@@ -27,202 +25,218 @@ func TestSend(t *testing.T) {
 		t.Fatal(err)
 	}
-	if receipt == nil {
-		t.Fatal("no receipt")
-	}
 
-	m, err := trojan.Unwrap(storedChunk)
+	topic1 := pss.NewTopic("topic-1")
+	topic2 := pss.NewTopic("topic-2")
+	topic3 := pss.NewTopic("topic-3")
+
+	topics := []pss.Topic{topic, topic1, topic2, topic3}
+	unwrapTopic, msg, err := pss.Unwrap(ctx, privkey, storedChunk, topics)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	if !bytes.Equal(m.Payload, payload) {
-		t.Fatalf("payload mismatch expected %v but is %v instead", m.Payload, payload)
+	if !bytes.Equal(msg, payload) {
+		t.Fatalf("message mismatch: expected %x, got %x", payload, msg)
 	}
 
-	if !bytes.Equal(m.Topic[:], topic[:]) {
-		t.Fatalf("topic mismatch expected %v but is %v instead", m.Topic, topic)
+	if !bytes.Equal(unwrapTopic[:], topic[:]) {
+		t.Fatalf("topic mismatch: expected %x, got %x", topic[:], unwrapTopic[:])
 	}
 }
 
+type topicMessage struct {
+	topic pss.Topic
+	msg   []byte
+}
+
 // TestDeliver verifies that registering a handler on pss for a given topic and then submitting a trojan chunk with said topic to it
 // results in the execution of the expected handler func
 func TestDeliver(t *testing.T) {
-	pss := pss.New(logging.New(ioutil.Discard, 0))
 	ctx := context.Background()
-	var mtx sync.Mutex
 
-	// test message
-	topic := trojan.NewTopic("footopic")
-	payload := []byte("foopayload")
-	msg, err := trojan.NewMessage(topic, payload)
+	privkey, err := crypto.GenerateSecp256k1Key()
 	if err != nil {
 		t.Fatal(err)
 	}
+	p := pss.New(privkey, logging.New(ioutil.Discard, 0))
+
+	target := pss.Target([]byte{1}) // arbitrary test target
+	targets := pss.Targets([]pss.Target{target})
+	payload := []byte("some payload")
+	topic := pss.NewTopic("topic")
+
+	recipient := &privkey.PublicKey
+
 	// test chunk
-	target := trojan.Target([]byte{1}) // arbitrary test target
-	targets := trojan.Targets([]trojan.Target{target})
-	c, err := msg.Wrap(ctx, targets)
+	chunk, err := pss.Wrap(context.Background(), topic, payload, recipient, targets)
 	if err != nil {
 		t.Fatal(err)
 	}
 
+	msgChan := make(chan topicMessage)
+
 	// create and register handler
-	var tt trojan.Topic // test variable to check handler func was correctly called
-	hndlr := func(ctx context.Context, m *trojan.Message) {
-		mtx.Lock()
-		copy(tt[:], m.Topic[:]) // copy the message topic to the test variable
-		mtx.Unlock()
+	handler := func(ctx context.Context, m []byte) {
+		msgChan <- topicMessage{
+			topic: topic,
+			msg:   m,
+		}
 	}
-	pss.Register(topic, hndlr)
+	p.Register(topic, handler)
 
 	// call pss TryUnwrap on chunk and verify test topic variable value changes
-	err = pss.TryUnwrap(ctx, c)
+	err = p.TryUnwrap(ctx, chunk)
 	if err != nil {
 		t.Fatal(err)
 	}
-	runtime.Gosched() // schedule the handler goroutine
-	for i := 0; i < 10; i++ {
-		mtx.Lock()
 
-		eq := bytes.Equal(tt[:], msg.Topic[:])
-		mtx.Unlock()
-		if eq {
-			return
-		}
-		<-time.After(50 * time.Millisecond)
+	message := <-msgChan
```

Comment:
Just not to wait for test to reach the global timeout, a select case with some reasonable large timeout would be nice.